save-preprocessor/src/main/kotlin/org/cqfn/save/preprocessor/controllers/DownloadProjectController.kt

```diff
@@ -107,6 +110,76 @@ class DownloadProjectController(private val configProperties: ConfigProperties)
                 .subscribe()
         }
 
+    /**
+     * Controller to download standard test suites
+     */
+    @Suppress("TOO_LONG_FUNCTION")
+    @PostMapping("/uploadStandardTestSuite")
+    fun uploadStandardTestSuite() {
+        readStandardTestSuitesFile().forEach { (testSuiteUrl, testSuitesPath) ->
+            log.info("Starting clone repository = $testSuiteUrl for standard test suites")
+            val tmpDir = generateDirectory(testSuiteUrl)
+            cloneFromGit(GitDto(testSuiteUrl), tmpDir).use {
+                Flux.fromIterable(testSuitesPath).flatMap { propPath ->
+                    log.info("Starting to discover root test config")
+                    val testResourcesRootAbsolutePath =
+                            File(
+                                tmpDir.relativeTo(File(configProperties.repository)).normalize().path,
+                                propPath
+                            ).absolutePath
+                    val rootTestConfig = testDiscoveringService.getRootTestConfig(testResourcesRootAbsolutePath)
+                    log.info("Starting to discover standard test suites")
+                    val tests = testDiscoveringService.getAllTestSuites(null, rootTestConfig, "stub")
+                    log.info("Test suites size = ${tests.size}")
+                    log.info("Starting to save new test suites")
+                    webClientBackend.makeRequest(BodyInserters.fromValue(tests), "/saveTestSuites") {
+                        it.bodyToMono<List<TestSuite>>()
+                    }
+                        .flatMap { testSuites ->
+                            log.info("Starting to save new tests")
+                            webClientBackend.makeRequest(
+                                BodyInserters.fromValue(
+                                    testDiscoveringService.getAllTests(
+                                        rootTestConfig,
+                                        testSuites
+                                    )
+                                ),
+                                "/initializeTests"
+                            ) { it.toBodilessEntity() }
+                        }
+                }
+                    .doOnError {
+                        log.error("Error to update test with url = $testSuiteUrl")
+                    }
+                    .subscribe()
+            }
+        }
+    }
+
+    private fun readStandardTestSuitesFile() =
+            ClassPathResource(configProperties.reposFileName)
+                .file
+                .readText()
+                .lines()
+                .associate {
+                    val splitRow = it.split("\\s".toRegex())
+                    require(splitRow.size == 2)
```

Comment:
If tests are located starting in repo root, can user specify `"."` as a location?

Reply:
I thought, that only we will specify this location, but anyway, it is possible.

Reply:
Yes, I meant someone who edits the file, so probably us :)